agent/xds/listeners.go

```diff
@@ -326,7 +327,50 @@ func (s *ResourceGenerator) listenersFromSnapshotConnectProxy(cfgSnap *proxycfg.
 		// Below we create a filter chain per upstream, rather than a listener per upstream
 		// as we do for explicit upstreams above.
 
-		// TODO(peering): tproxy
+		filterChain, err := s.makeUpstreamFilterChain(filterChainOpts{
+			routeName:   uid.EnvoyID(),
+			clusterName: clusterName,
+			filterName:  uid.EnvoyID(),
+			protocol:    cfg.Protocol,
+			useRDS:      false,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		endpoints, _ := cfgSnap.ConnectProxy.PeerUpstreamEndpoints.Get(uid)
+		uniqueAddrs := make(map[string]struct{})
+
+		// Match on the virtual IP for the upstream service (identified by the chain's ID).
+		// We do not match on all endpoints here since it would lead to load balancing across
+		// all instances when any instance address is dialed.
+		for _, e := range endpoints {
+			if vip := e.Service.TaggedAddresses[structs.TaggedAddressVirtualIP]; vip.Address != "" {
+				uniqueAddrs[vip.Address] = struct{}{}
+			}
+
+			// The virtualIPTag is used by consul-k8s to store the ClusterIP for a service.
+			// For services imported from a peer,the partition will be equal in all cases.
+			if acl.EqualPartitions(e.Node.PartitionOrDefault(), cfgSnap.ProxyID.PartitionOrDefault()) {
+				if vip := e.Service.TaggedAddresses[virtualIPTag]; vip.Address != "" {
```

Comment:
Maybe @freddygv will have some context? I poked around at the history and it looks like the `virtualIPTag` was for initial K8s support of TProxy before Consul was later updated to use `structs.TaggedAddressVirtualIP` and assign VIPs directly. Some tests assign different IPs to the two keys but I'm not sure if that's valid

Reply:
* `virtual` is for the Kubernetes-assigned ClusterIP for services. As Chris mentioned this is the address that's expected to be used for cluster-local communication, since it holds the address returned by kubedns queries.
* `consul-virtual` is for the one assigned by Consul itself, it was introduced for cross-partition communication in 1.11, and is also used for addressing peered services. This requires configuring/using Consul DNS.
* `consul-virtual:<service-name>` is a variant attached to terminating gateways. Since a gateway endpoint can route to multiple services, there is a virtual IP attached to every gateway instance for each of those linked service names.